src/astro/sql/operators/cleanup.py

```diff
@@ -0,0 +1,164 @@
+    tasks: List[BaseOperator], context: Context
+) -> List[Table]:
+    """
+    For the moment, these are the only two classes that create temporary tables.
+    This function allows us to only resolve xcom for those objects (to reduce how much data is brought into the worker).
+
+    We also process these values one at a time so the system can garbage collect non-table objects (otherwise we might
+    run into a situation where we pull in a bunch of dataframes and overwhelm the worker).
+    :param tasks: A list of operators from airflow that we can resolve
+    :param context: Context of the DAGRun so we can resolve against the XCOM table
+    :return:
+    """
+    res = []
+    for task in tasks:
+        if isinstance(task, (DataframeOperator, BaseSQLOperator, LoadFile)):
+            t = task.output.resolve(context)
+            if isinstance(t, Table):
+                res.append(t)
+    return res
+
+
+def filter_for_temp_tables(task_outputs: List[Any]) -> List[Table]:
+    return [t for t in task_outputs if isinstance(t, Table) and t.temp]
+
+
+class CleanupOperator(BaseOperator):
+    """
+    Clean up temporary tables at the end of a DAG run.
+
+    By default if no tables are placed, the task will wait for all other tasks to run before deleting all temporary
+    tables (WARNING: DO NOT RUN THIS MODE IF USING THE SEQUENTIAL EXECUTOR. IT WILL CAUSE A BLOCK).
+    :param tables_to_cleanup: List of tbles to drop at the end of the DAG run
+    :param task_id: Optional custom task id
+    :param run_sync_mode: Whether to wait for the DAG to finish or not. Set to False if you want to immediately
+    clean all DAGs. Note that if you supply anything to `tables_to_cleanup` this argument is ignored.
+    """
+
+    template_fields = ("tables_to_cleanup",)
+
+    def __init__(
+        self,
+        *,
+        tables_to_cleanup: Optional[List[Table]] = None,
+        task_id: str = "",
+        run_sync_mode: bool = False,
+        **kwargs,
+    ):
+        self.tables_to_cleanup = tables_to_cleanup or []
+        self.run_immediately = run_sync_mode
+        self.single_worker_mode = os.getenv("AIRFLOW__ASTRO__SINGLE_WORKER_MODE")
+        task_id = task_id or get_unique_task_id("_cleanup")
+
+        super().__init__(task_id=task_id, **kwargs)
+
+    def execute(self, context: Context):
+        self.log.info("Execute Cleanup")
+        if not self.tables_to_cleanup:
+            # tables not provided, attempt to either immediately run or wait for all other tasks to finish
+            if not self.run_immediately:
+                self.wait_for_dag_to_finish(context)
+            self.tables_to_cleanup = self.get_all_task_outputs(context=context)
+        self.log.info(
+            "Tables found for cleanup: %s",
+            ",".join([t.name for t in self.tables_to_cleanup]),
+        )
+        temp_tables = filter_for_temp_tables(self.tables_to_cleanup)
+        for table in temp_tables:
+            self.drop_table(table)
```

Comment:
@dimberman  I think we should make this fault-tolerant if there is an error due to network fault, the subsequent tables deletion will be skipped.

Reply:
@utkarsharma2 I don't think we need to worry about that as there are two incoming fixes that will make that irrelevant.

1. @sunank200 is going to make this operator asynchronous
2. In airflow 2.4 we will have DAG-level teardown operations so this won't even be necessary anymore.